### PR TITLE
Make SCAN faster.

### PR DESCRIPTION
**Pull request by jpountz**
When scrolling, SCAN previously collected documents until it reached where it
had stopped on the previous iteration. This makes pagination slower and slower
as you request deep pages. With this change, SCAN now directly jumps to the
doc ID where is had previously stopped.

### CODE REVIEW

File: src/main/java/org/elasticsearch/search/scan/ScanContext.java [hunk 1]
```diff
@@ -112,70 +88,113 @@ public void setScorer(Scorer scorer) throws IOException {
...
+            docBase = context.docBase;
         }
     }
 
-    public static class ScanFilter extends Filter {
+    /**
+     * A filtering query that matches all doc IDs that are not deleted and
+     * greater than or equal to the configured doc ID.
+     */
+    private static class MinDocQuery extends Query {
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/search/scan/ScanContext.java]: neat!

File: src/main/java/org/elasticsearch/search/scan/ScanContext.java [hunk 2]
```diff
@@ -112,70 +88,113 @@ public void setScorer(Scorer scorer) throws IOException {
...
+                        }
+
+                        @Override
+                        public int nextDoc() throws IOException {
+                            return advance(doc + 1);
+                        }
+
+                        @Override
+                        public int advance(int target) throws IOException {
+                            assert target > doc;
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/search/scan/ScanContext.java]: would be great if this logic could be unit tested.

**Comment by martijnvg** [comment]
I like this approach! Left one minor comment, but other than that LGTM.

**Comment by jpountz** [comment]
@martijnvg I pushed a new commit.

**Comment by martijnvg** [comment]
LGTM

**Comment by clintongormley** [comment]
wow!

**Pull request merged**